src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseMetaService.cs

```diff
@@ -0,0 +1,74 @@
+using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Common.Utils;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Data.Model;
+using GovUk.Education.ExploreEducationStatistics.Data.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.ViewModels.Meta;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Release = GovUk.Education.ExploreEducationStatistics.Content.Model.Release;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
+{
+    public class ReleaseMetaService : IReleaseMetaService
+    {
+        private readonly ContentDbContext _contentDbContext;
+        private readonly IPersistenceHelper<ContentDbContext> _contentPersistenceHelper;
+        private readonly StatisticsDbContext _statisticsDbContext;
+        private readonly IUserService _userService;
+
+        public ReleaseMetaService(ContentDbContext contentDbContext,
+            IPersistenceHelper<ContentDbContext> contentPersistenceHelper,
+            StatisticsDbContext statisticsDbContext,
+            IUserService userService)
+        {
+            _contentDbContext = contentDbContext;
+            _contentPersistenceHelper = contentPersistenceHelper;
+            _statisticsDbContext = statisticsDbContext;
+            _userService = userService;
+        }
+
+        public async Task<Either<ActionResult, SubjectsMetaViewModel>> GetSubjects(Guid releaseId)
+        {
+            return await _contentPersistenceHelper
+                .CheckEntityExists<Release>(releaseId)
+                .OnSuccess(_userService.CheckCanViewRelease)
+                .OnSuccess(release =>
+                {
+                    var files = _contentDbContext.ReleaseFiles
+                        .Include(file => file.ReleaseFileReference)
+                        .Where(file => file.ReleaseId == releaseId
+                                       && file.ReleaseFileReference.ReleaseFileType == ReleaseFileTypes.Data)
+                        .Select(file => file.ReleaseFileReference);
+
+                    // Exclude files that are replacements in progress
+                    var filesExcludingReplacements =
+                        files.Where(file => !file.ReplacingId.HasValue);
+
+                    var subjectIds = filesExcludingReplacements.Select(file => file.SubjectId).ToList();
+
+                    var subjects = _statisticsDbContext.ReleaseSubject
+                        .Include(subject => subject.Subject)
+                        .Where(subject => subject.ReleaseId == releaseId
+                                          && subjectIds.Contains(subject.SubjectId))
```

Comment:
Rather than use a contains with a bunch of ids, wonder if it's possible to use a manual `Join`?

Reply:
We can't use multiple database contexts in a single query so we'd have to at least turn one side of the join into a List in memory. That's ok here as `subjectIds` is already a List or we could get rid of it and `filesExcludingReplacements` can be a List.

However, even this isn't possible now since EF Core 3.

https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-3.x/breaking-changes#linq-queries-are-no-longer-evaluated-on-the-client

> Starting with 3.0, EF Core only allows expressions in the top-level projection (the last Select() call in the query) to be evaluated on the client. When expressions in any other part of the query can't be converted to either SQL or a parameter, an exception is thrown.

```
                    var subjects = _statisticsDbContext.ReleaseSubject
                        .Join(filesExcludingReplacements,
                            rs => rs.SubjectId,
                            rfr => rfr.SubjectId.Value,
                            (rs, _) => new IdLabel(rs.Subject.Id, rs.Subject.Name))
                        .ToList();
```

gives error

```
System.InvalidOperationException : Processing of the LINQ expression 'DbSet<ReleaseSubject>
    .Join(
        outer: __p_0, 
        inner: rs => rs.SubjectId, 
        outerKeySelector: rfr => rfr.SubjectId.Value, 
        innerKeySelector: (rs, _) => new IdLabel(
            rs.Subject.Id, 
            rs.Subject.Name
        ))' by 'NavigationExpandingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core.
```

when using `filesExcludingReplacements` as a List or the same error when using the list of `subjectIds`.

So the only way to get this join to work would be to enumerate both sides as Lists in memory first and then do a join.

Doing `_statisticsDbContext.ReleaseSubject.ToList()` would be crazy, and is probably why they restricted this if that's what it was doing internally before.

We could do this but I'm not sure it's any better than what we had before

```
                    var filesExcludingReplacements =
                        files.Where(file => !file.ReplacingId.HasValue).ToList();

                    var releaseSubjects = 
                        _statisticsDbContext.ReleaseSubject
                            .Include(rs => rs.Subject)
                            .Where(subject => subject.ReleaseId == releaseId).ToList();

                    var subjects = releaseSubjects
                        .Join(filesExcludingReplacements,
                            rs => rs.SubjectId,
                            rfr => rfr.SubjectId,
                            (rs, _) => new IdLabel(rs.Subject.Id, rs.Subject.Name))
                        .ToList();
```

Reply:
Cool, that makes sense. Was a bit optimistic to think it would be able to!